### DISCUSSION TITLE
Cached property getter

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5180] [Language Ideas]
I've often implemented property getters in the pattern:
```csharp
T Property
{
    get
    {
        if (field != null)
            return field;

        var result = ...; // some operation to get the property value
        return field = result;
    }
}
```

Where `field` is the backing field, and `result` stores the returned result of the operation.

I would like this to be simplified down to something like:
```csharp
T Property
{
    init get
    {
        return ...; // just the operation itself
    }
}
```

`init get` is a special-cased property accessor, that initializes the value of the property upon being called for the first time, and is returned after being initialized. `init get` accessors are only valid in non-auto-implemented properties (`{ init get; }` is invalid), and no other accessors can be implemented (`init get` cannot be paired with any other accessor).

## Unresolved Questions
- Can accessor properties (`this[...]`) have an `init get` accessor?

**Comment by FaustVX** [comment]
You could use the `null coalescing operator` (`??=`) :

```csharp
private string _p = null;
public string P => _p ??= "Value";
```

**Reply by Rekkonnect** [reply]
This is not quite convenient for cases where the operation is a string of operations, including invocations, member accesses, etc. For simpler cases, this is what I've been following, and in C# 10 it will be improved even further with the `field` keyword, but there is no good solution if you cannot smudge the whole thing into one line.

**Reply by FaustVX** [reply]
With C#10, you could do :
```csharp
public string P => field ??= GetP(); // your one-liner

private string GetP()
{
  … // All computation goes here
}
```

**Reply by Rekkonnect** [reply]
Does that not pretty much eliminate the point of properties? From what I can tell, properties are meant to be syntax sugar to encapsulate a getter and a setter for a value. This is basically going back to those primal days.

I do recognize that it is a workaround, but the need for the feature is not eliminated through its usage.

**Reply by FaustVX** [reply]
Usually, you don't have much logic in the getter.
Everything goes in the setter or directly in the field.

**Reply by jnm2** [reply]
@AlFasGD @FaustVX Unfortunately, `field` (#140) is definitely not coming in C# 10. This is what 10 will have: https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md#c-100

**Reply by jnm2** [reply]
If I need multiple statements, I usually do either a `CreateP()` method similar to the above, or this:

```cs
private readonly Lazy<string> p = new(() =>
{
    // Multiple statements
});

public string P => p.Value;
```

**Reply by FaustVX** [reply]
Ok, too bad for `field` (there is no release version yet ?)
Yes, I always forget about `Lazy<>`

**Reply by jnm2** [reply]
No release version yet. The latest news that has been shared is at https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-04-14.md#field-keyword, and maybe @CyrusNajmabadi can confirm expectations.

**Reply by jnm2** [reply]
Oh, missed this one in the "possibilities for C# 11" list: https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-08-30.md#field-keyword

**Reply by FaustVX** [reply]
@jnm2  Ok, thanks
